wagtail/wagtailadmin/views/pages.py

```diff
@@ -581,12 +582,23 @@ def get_form(self):
         page = self.get_page()
         form_class = page.get_edit_handler().get_form_class(page._meta.model)
         parent_page = page.get_parent().specific
-        post_data, timestamp = self.request.session[self.session_key]
+        post_data_dict, timestamp = self.request.session[self.session_key]
+
+        # convert post_data_dict back into a QueryDict
+        post_data = QueryDict('', mutable=True)
+        for k, v in post_data_dict.items():
+            post_data.setlist(k, v)
+
         return form_class(post_data, instance=page, parent_page=parent_page)
 
     def post(self, request, *args, **kwargs):
         # TODO: Handle request.FILES.
-        request.session[self.session_key] = request.POST, time()
+
+        # Convert request.POST to a plain dict (rather than a QueryDict) so that it can be
+        # stored without data loss in session data
+        post_data_dict = dict(request.POST.lists())
```

Comment:
It could be simpler by converting to tuple, then remove the `.items()` line 589.
An even simpler solution would be to convert to string using `QueryDict.urlencode()`, then directly initialise the `QueryDict` with this string when restoring data. Maybe I miss something, but it should work and it would be more compact in memory and faster to serialize/deserialize in session data.